mmdet3d/models/losses/lovasz_loss.py

```diff
@@ -0,0 +1,398 @@
+        reduction (str): The method used to reduce the loss. Options
+            are "none", "mean" and "sum". This parameter only works when
+            per_sample is True. Defaults to 'mean'.
+        avg_factor (int, optional): Average factor that is used to average
+            the loss. This parameter only works when per_sample is True.
+            Defaults to None.
+        ignore_index (Union[int, None]): The label index to be ignored.
+            Defaults to 255.
+
+    Returns:
+        torch.Tensor: The calculated loss.
+    """
+
+    if per_sample:
+        loss = [
+            lovasz_softmax_flat(
+                *flatten_probs(
+                    prob.unsqueeze(0), label.unsqueeze(0), ignore_index),
+                classes=classes,
+                class_weight=class_weight)
+            for prob, label in zip(probs, labels)
+        ]
+        loss = weight_reduce_loss(
+            torch.stack(loss), None, reduction, avg_factor)
+    else:
+        loss = lovasz_softmax_flat(
+            *flatten_probs(probs, labels, ignore_index),
+            classes=classes,
+            class_weight=class_weight)
+    return loss
+
+
+@MODELS.register_module()
+class LovaszLoss(nn.Module):
+    """LovaszLoss.
+
+    This loss is proposed in `The Lovasz-Softmax loss: A tractable surrogate
+    for the optimization of the intersection-over-union measure in neural
+    networks <https://arxiv.org/abs/1705.08790>`_.
+
+    Args:
+        loss_type (str): Binary or multi-class loss.
+            Defaults to 'multi_class'. Options are "binary" and "multi_class".
+        classes (Union[str, list[int]]): Classes chosen to calculate loss.
+            'all' for all classes, 'present' for classes present in labels, or
+            a list of classes to average. Defaults to 'present'.
+        per_sample (bool): If per_sample is True, compute the loss per
+            sample instead of per batch. Defaults to False.
+        reduction (str): The method used to reduce the loss. Options
+            are "none", "mean" and "sum". This parameter only works when
+            per_sample is True. Defaults to 'mean'.
+        class_weight (Union[list[float], str], optional): Weight of each class.
+            If in str format, read them from a file. Defaults to None.
+        loss_weight (float): Weight of the loss. Defaults to 1.0.
+        loss_name (str): Name of the loss item. If you want this loss
+            item to be included into the backward graph, `loss_` must be the
+            prefix of the name. Defaults to 'loss_lovasz'.
+    """
+
+    def __init__(self,
+                 loss_type: str = 'multi_class',
+                 classes: Union[str, List[int]] = 'present',
+                 per_sample: bool = False,
+                 reduction: str = 'mean',
+                 class_weight: Optional[Union[List[float], str]] = None,
+                 loss_weight: float = 1.0,
+                 loss_name: str = 'loss_lovasz'):
```

Comment:
In mmdet3d, do we need loss_name arguments?

Reply:
I remove it.